lesson11/main.js

```diff
@@ -0,0 +1,69 @@
+const ul = document.getElementById("js-ul");
+
+const addLoading = () =>{
+  const loading = document.getElementById("js-loading");
+  const loadingImg = document.createElement('img');
+  loadingImg.src = 'loading.gif';
+  loadingImg.id = "loading-img";
+  loading.appendChild(loadingImg);
+};
+
+const removeLoading = () =>{
+  document.getElementById('loading-img').remove();
+};
+
+const renderLists = (features) =>{
+  const fragment = new DocumentFragment();
+
+  for (let feature of features ){
+      const li = document.createElement('li');
+      const a = document.createElement('a');
+      const img = document.createElement('img');
+      a.href = feature.to;
+      a.textContent = feature.text;
+      img.src = feature.img;
+      img.alt = feature.alt;
+
+      fragment.appendChild(li).appendChild(a).insertAdjacentElement('afterbegin', img);
+    };
+
+    ul.appendChild(fragment);
+};
+
+const displayError = (error) =>{
+  const li = document.createElement('li');
+  li.textContent = error;
+  ul.appendChild(li);
+};
+
+
+
+const checkResponseStatus = async () => {
+  const fetchedData = await fetch('https://mocki.io/v1/ee8a871e-2b46-4a91-b565-4d6f9216f300');
+  if(fetchedData.status === 200){
+    return fetchedData.json();
+  } else {
+    return Promise.reject(new Error('Failed to fetch data from server!'));
+  }
```

Comment:
IMO
With early returns, you can do it in two lines.
[The early return pattern in JavaScript](https://gomakethings.com/the-early-return-pattern-in-javascript/)
```suggestion
  if (fetchedData.status === 200) return fetchedData.json();
  return Promise.reject(new Error("Failed to fetch data from server!"));
```

Reply:
IMO
In the Fetch API, response.ok can be used to tell if the response was successful (status in the 200-299 range) or not. Together with Yoshimi san's code, I think you can write the following.
```JavaScript
if(fetchedData.ok) return fetchedData.json();
return Promise.reject(new Error("Failed to fetch data from server!"));
```

Reply:
Thank you for your suggestion Yoshimi-san and Mai-san,

I changed them as per suggested 
Please see the latest commit 
https://github.com/Saki-Ota/JS_practice/commit/96057c52090197748e4cabff8d120d3c726ddf63

Reply:
> Please see the latest commit
96057c5

Q
Is it possible to change it to two lines since it is on one line?

Reply:
Q
I also wanted to know why you use Promise.reject. If you have a reason, please let me know.
I didn't think it was necessary to write Promise.reject because fetch does resolve and reject internally on its own.
The fecth API is handled as a reject in the event of a server error.

Reply:
Mai-san,
In my understanding, the promise returned by fetch() is fulfilled when it finishes passing the value  and it does not returned as reject even if it did not get data from server. If try/catch method was used within the same function, reject value is not needed to be defined, however my original code passes the returned value of function checkResponseStatus() to displayListsOrError(). When the promise is passed to another function, it will be passed as fulfilled regardless it was successful or not. Therefore, the reject value was defined in the function checkResponseStatus() so that when fetch() was failed it will pass reject value to displayListsOrError().

But as Morita-san explained, try/catch should be used right after fetch to check if it successfully fetches data from server or not, in this case reject value does not have to be defined. 
Hoep it makes sense


Reference 
https://kde.hateblo.jp/entry/2018/10/22/010811

Reply:
Thank you! So you are taking all server responses as resolve, right? 
Now I understand! 
Thanks for the detailed explanation. 
Thanks for giving me a chance to understand Promise in depth.